linen_examples/pixelcnn/model_test.py

```diff
@@ -0,0 +1,134 @@
+    self.x = np.arange(24).reshape(1, 4, 3, 2)
+
+
+  def get_weightnorm(self, params):
+    return [params[k] for k in ('direction', 'scale', 'bias')]
+
+
+  def assert_mean_and_variance(self, out):
+    # Weightnorm should ensure that, at initialization time, the outputs of the
+    # module have mean 0 and variance 1 over the non-feature dimensions.
+    onp_testing.assert_allclose(np.mean(out, (0, 1, 2)), 0., atol=1e-5)
+    onp_testing.assert_allclose(np.var(out, (0, 1, 2)), 1., atol=1e-5)
+
+
+  def test_conv(self):
+    model = pixelcnn.ConvWeightNorm(features=4, kernel_size=(3, 2))
+    out, variables = model.init_with_output(self.rng, self.x)
+    params = variables['param']['weightnorm_params']
+    direction, scale, bias = self.get_weightnorm(params)
+
+    self.assertEqual(direction.shape, (3, 2, 2, 4))
+    self.assertEqual(scale.shape, (4,))
+    self.assertEqual(bias.shape, (4,))
+    self.assertEqual(out.shape, (1, 2, 2, 4))
+    self.assert_mean_and_variance(out)
+
+
+  def test_conv_down(self):
+    model = pixelcnn.ConvDown(features=4)
+    out, variables = model.init_with_output(self.rng, self.x)
+    params = variables['param']['ConvWeightNorm_0']['weightnorm_params']
+    direction, scale, bias = self.get_weightnorm(params)
+
+    self.assertEqual(direction.shape, (2, 3, 2, 4))
+    self.assertEqual(scale.shape, (4,))
+    self.assertEqual(bias.shape, (4,))
+    self.assertEqual(out.shape, (1, 4, 3, 4))
+    self.assert_mean_and_variance(out)
+
+
+  def test_conv_down_right(self):
+    model = pixelcnn.ConvDownRight(features=4)
+    out, variables = model.init_with_output(self.rng, self.x)
+    params = variables['param']['ConvWeightNorm_0']['weightnorm_params']
+    direction, scale, bias = self.get_weightnorm(params)
+
+    self.assertEqual(direction.shape, (2, 2, 2, 4))
+    self.assertEqual(scale.shape, (4,))
+    self.assertEqual(bias.shape, (4,))
+    self.assertEqual(out.shape, (1, 4, 3, 4))
+    self.assert_mean_and_variance(out)
+
+
+  def test_conv_transpose(self):
+    model = pixelcnn.ConvTranspose(features=4, kernel_size = (3, 2))
+    out, variables = model.init_with_output(self.rng, self.x)
+    params = variables['param']['weightnorm_params']
+    direction, scale, bias = self.get_weightnorm(params)
+
+    self.assertEqual(direction.shape, (3, 2, 2, 4))
+    self.assertEqual(scale.shape, (4,))
+    self.assertEqual(bias.shape, (4,))
+    self.assertEqual(out.shape, (1, 6, 4, 4))
+    self.assert_mean_and_variance(out)
+    
+
+  def test_conv_transpose_down(self):
+    model = pixelcnn.ConvTransposeDown(features=4)
+    out, variables = model.init_with_output(self.rng, self.x)
+    params = variables["param"]["ConvWeightNorm_0"]["weightnorm_params"]
+    direction, scale, bias = self.get_weightnorm(params)
+
+    self.assertEqual(direction.shape, (2, 3, 2, 4))
+    self.assertEqual(scale.shape, (4,))
+    self.assertEqual(bias.shape, (4,))
+    self.assertEqual(out.shape, (1, 8, 6, 4))
```

Comment:
Looks like you accidentally removed the test for actual mean and stddev here?

Reply:
Actually it wasn't meant to be in the test, so remove it is correct.